src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java

```diff
@@ -51,14 +93,37 @@
                             GOCARDLESS_ACTIONS_THAT_CHANGE_STATE);
                 })
                 .orElse(Optional.empty());
+    }
+
+    private DirectDebitStateWithDetails<MandateState> mapEventToState(Event event, Mandate mandate) {
+        if (event instanceof GoCardlessEvent) {
+            return mapGoCardlessEventToState((GoCardlessEvent) event, mandate);
+        } else if (event instanceof GovUkPayEvent) {
+            return mapGovUkPayEventToState((GovUkPayEvent) event, mandate);
+        } else {
+            throw new IllegalArgumentException(format("Unexpected Event of type %s", event.getClass()));
+        }
+    }
+
+    private DirectDebitStateWithDetails<MandateState> mapGoCardlessEventToState(GoCardlessEvent goCardlessEvent, Mandate mandate) {
+        if (!GOCARDLESS_ACTION_TO_MANDATE_STATE.containsKey(goCardlessEvent.getAction())) {
+            throw new UnexpectedGoCardlessEventActionException(goCardlessEvent.getAction(), mandate);
+        }
+        return new DirectDebitStateWithDetails<>(
+                GOCARDLESS_ACTION_TO_MANDATE_STATE.get(goCardlessEvent.getAction()),
+                goCardlessEvent.getDetailsCause(),
+                goCardlessEvent.getDetailsDescription());
+    }
 
-        return latestApplicableGoCardlessEvent
-                .filter(goCardlessEvent -> GOCARDLESS_ACTION_TO_MANDATE_STATE.get(goCardlessEvent.getAction()) != null)
-                .map(goCardlessEvent -> new DirectDebitStateWithDetails<>(
-                        GOCARDLESS_ACTION_TO_MANDATE_STATE.get(goCardlessEvent.getAction()),
-                        goCardlessEvent.getDetailsCause(),
-                        goCardlessEvent.getDetailsDescription())
-                );
+    private DirectDebitStateWithDetails<MandateState> mapGovUkPayEventToState(GovUkPayEvent govUkPayEvent, Mandate mandate) {
+        if (!GOV_UK_PAY_EVENT_TYPE_MANDATE_STATE.containsKey(govUkPayEvent.getEventType())) {
+            throw new UnexpectedGovUkPayEventTypeException(govUkPayEvent.getEventType(), mandate);
+        }
+        return new DirectDebitStateWithDetails<>(
+                GOV_UK_PAY_EVENT_TYPE_MANDATE_STATE.get(govUkPayEvent.getEventType()),
+                null,
+                null
+        );
```

Comment:
I wonder if it’s time for an overloaded constructor

Reply:
ok